File: Java/src/main/java/at/tugraz/networkadapter/Network.java
package at.tugraz.networkadapter;

import org.apache.log4j.Logger;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.net.InetAddress;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;

public class Network {
    private Logger logger = Logger.getLogger(Network.class);
    private HashMap<String, MachineInfo> machineInfos;
    String config;

    public Network(String config) throws IOException {
        this.config = config;
        this.readConfigFile(config);
    }

    public void readConfigFile(String config) throws IOException {
        machineInfos = new HashMap<>();
        Map<String, String> rootLeaf = new HashMap<>();

        try (BufferedReader br = new BufferedReader(new FileReader(config, StandardCharsets.UTF_8))) {
            String line;
            br.readLine();
            while ((line = br.readLine()) != null) {
                String[] cols = line.split(",");
                String ip = cols[0];
                int port = Integer.parseInt(cols[1]);
                String root = cols[2];
                int nrow = Integer.parseInt(cols[3]);
                rootLeaf.put(ip, root);
                MachineInfo nsi = new MachineInfo(ip, port, nrow);
                machineInfos.put(ip, nsi);
            }
        }

        for (String ip: rootLeaf.keySet()){
            String root = rootLeaf.get(ip);
            if (root.equalsIgnoreCase("root"))
                continue;
            machineInfos.get(root).addLeaf(machineInfos.get(ip));
            machineInfos.get(ip).setRoot(machineInfos.get(root));
        }

        for (String ip: machineInfos.keySet()){
            int sumLeaf = computeNRow(machineInfos.get(ip)) - machineInfos.get(ip).getNrow();
            machineInfos.get(ip).setNrowLeaf(sumLeaf);
        }
    }

    public MachineInfo getCurrentMachine() {
        try {
            String ipstr;
            InetAddress ip;
            ip = InetAddress.getLocalHost();
            ipstr = ip.getHostAddress();
            return machineInfos.get(ipstr);

        } catch (Exception ignored) {
            logger.error("Get Current Machine IP:" + ignored);
        }
        throw new RuntimeException("The machine IP address is not in the config list");
    }

    private int computeNRow(MachineInfo node){
        if (node.getLeaves().size() == 0)
            return node.getNrow();
        else {
            int sum = node.getNrow();
            for (MachineInfo nsi: node.getLeaves())
                sum += computeNRow(nsi);
            return sum;
        }
    }
}

File: src/main/rust/src/util/Network.rs
use std::borrow::{Borrow, BorrowMut};
use std::collections::HashMap;
use crate::util::MachineInfo::MachineInfo;
use std::error::Error;
use std::io;
use std::process;
use csv;
use csv::Reader;
use local_ip_address::local_ip;

pub struct Network {
    config: String,
    machineInfo: HashMap<String, MachineInfo>,
}

impl Network {
    pub fn new(config: String) -> Self {
        let list = Network::readConfigFile(config.clone());
        Self {
            config: config,
            machineInfo: list,
        }
    }

    fn readConfigFile(config: String) -> HashMap<String, MachineInfo> { //-> HashMap<String, MachineInfo>
        let mut list: HashMap<String, MachineInfo> = HashMap::new();
        let mut rootLeaf: HashMap<String, String> = HashMap::new();

        let mut br = csv::Reader::from_path(config.clone()).unwrap();
        for line in br.records() {
            let record = line.unwrap();
            let ip = record.get(0).unwrap().to_string();
            let port: u16 = record.get(1).unwrap().parse().unwrap();
            let root = record.get(2).unwrap().to_string();
            let nrow: u32 = record.get(3).unwrap().parse().unwrap();
            rootLeaf.insert(ip.clone(), root.clone());

            let nsi = MachineInfo::new(ip.clone(), port, root, nrow);
            list.insert(ip, nsi);

        }

        for ip in rootLeaf.keys() {
            let root = rootLeaf.get(ip.clone().as_str()).unwrap();
            if root.to_lowercase().eq("root") {
                continue;
            }
            list.get_mut(root.clone().as_str()).unwrap().addLeaf(ip.clone());
            list.get_mut(ip.clone().as_str()).unwrap().set_root(root.clone());
        }

        for ip in rootLeaf.keys() {
            let sumLeaf = Network::computeNRow(&list, ip);
            list.get_mut(ip).unwrap().set_nrowLeaf(sumLeaf);
        }
        return list;
    }

    fn computeNRow(list: &HashMap<String, MachineInfo>, node: &str) -> u32 {
        let tmp = list.get(node).borrow().unwrap();
        if tmp.leaves().len() == 0 {
            return tmp.nrow();
        } else {
            let mut sum = tmp.nrow();
            for leaf in tmp.leaves() {
                sum += Network::computeNRow(list.borrow(), leaf);
            }
            return sum;
        }
    }

    pub fn getCurrentMachine(&mut self) -> &mut MachineInfo {
        let ipstr = local_ip().unwrap().to_string();
        return self.machineInfo.get_mut(ipstr.as_str()).unwrap();
    }
}